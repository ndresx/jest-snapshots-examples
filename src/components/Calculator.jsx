import React from 'react';

export function sum(a, b) {
  return a + b;
}

// Custom button component
const Button = ({ onClick, children }) => <button onClick={onClick}>{children}</button>;

const defaultState = { input1: '', input2: '', output: null };

export default class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = { ...defaultState };
  }

  handleInput1Change = e => {
    this.setState({ input1: parseFloat(e.target.value) || '' });
  };

  handleInput2Change = e => {
    this.setState({ input2: parseFloat(e.target.value) || '' });
  };

  handleAddBtnClick = () => {
    this.setState({
      output: sum(parseFloat(this.state.input1), parseFloat(this.state.input2)),
    });
  };

  handleResetBtnClick = () => {
    this.setState(defaultState);
  };

  renderOutput() {
    const { output } = this.state;

    if (output === null) {
      return null;
    }

    if (isNaN(output)) {
      return 'Please verify your input!';
    }

    return <div>Result: {output}</div>;
  }

  render() {
    return (
      <div className="calculator">
        <h1>Calculator</h1>
        <div>
          <label htmlFor="input-1">Input 1:</label>
          <input
            id="input-1"
            placeholder="Input 1"
            ref="input1"
            value={this.state.input1}
            onChange={this.handleInput1Change}
          />
        </div>
        <div>
          <label htmlFor="input-2">Input 2:</label>
          <input
            id="input-2"
            placeholder="Input 2"
            ref="input2"
            value={this.state.input2}
            onChange={this.handleInput2Change}
          />
        </div>
        <div id="output">{this.renderOutput()}</div>
        <button onClick={this.handleAddBtnClick}>Add</button>
        <Button onClick={this.handleResetBtnClick}>Reset</Button>
      </div>
    );
  }
}
