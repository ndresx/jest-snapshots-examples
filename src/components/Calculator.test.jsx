import React from 'react';
import Calculator, { sum } from './Calculator';
import { shallow } from 'enzyme';

it('should render correctly', () => {
  const calculator = shallow(<Calculator />);
  expect(calculator).toMatchSnapshot();
});

it('should render correctly (without snapshot)', () => {
  const calculator = shallow(<Calculator />);
  expect(calculator.find('h1').text()).toBe('Calculator');

  const inputs = calculator.find('input');
  expect(inputs.at(0).prop('id')).toBe('input-1');
  expect(inputs.at(1).prop('id')).toBe('input-2');

  expect(calculator.find('button').text()).toBe('Add');
  expect(calculator.find('Button').prop('children')).toBe('Reset');

  expect(calculator.find('#output').length).toBe(1);
});

it('should render error message if input is not valid', () => {
  const calculator = shallow(<Calculator />);

  calculator.find('button').simulate('click');
  const output = calculator.find('#output');

  // very bad, because it captures the full output again
  expect(calculator).toMatchSnapshot();

  // good, because self-descriptive, but requires text maintenance
  expect(output.text()).toBe('Please verify your input!');

  // better, because it only captures the result text without having to write the text
  expect(output.text()).toMatchSnapshot();

  // better, because it only captures the result text without having to actually write the text
  expect(output.text()).toMatchInlineSnapshot(`"Please verify your input!"`);
});

it('should return 4 when calculating 1 + 3', () => {
  expect(sum(1, 3)).toMatchSnapshot();
});

it('should return 4 when calculating 1 + 3 (without snapshot)', () => {
  expect(sum(1, 3)).toBe(4);
});
