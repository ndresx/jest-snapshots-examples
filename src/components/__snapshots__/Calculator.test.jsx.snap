// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="calculator"
>
  <h1>
    Calculator
  </h1>
  <div>
    <label
      htmlFor="input-1"
    >
      Input 1:
    </label>
    <input
      id="input-1"
      onChange={[Function]}
      placeholder="Input 1"
      value=""
    />
  </div>
  <div>
    <label
      htmlFor="input-2"
    >
      Input 2:
    </label>
    <input
      id="input-2"
      onChange={[Function]}
      placeholder="Input 2"
      value=""
    />
  </div>
  <div
    id="output"
  />
  <button
    onClick={[Function]}
  >
    Add
  </button>
  <Button
    onClick={[Function]}
  >
    Reset
  </Button>
</div>
`;

exports[`should render error message if input is not valid 1`] = `
<div
  className="calculator"
>
  <h1>
    Calculator
  </h1>
  <div>
    <label
      htmlFor="input-1"
    >
      Input 1:
    </label>
    <input
      id="input-1"
      onChange={[Function]}
      placeholder="Input 1"
      value=""
    />
  </div>
  <div>
    <label
      htmlFor="input-2"
    >
      Input 2:
    </label>
    <input
      id="input-2"
      onChange={[Function]}
      placeholder="Input 2"
      value=""
    />
  </div>
  <div
    id="output"
  >
    Please verify your input!
  </div>
  <button
    onClick={[Function]}
  >
    Add
  </button>
  <Button
    onClick={[Function]}
  >
    Reset
  </Button>
</div>
`;

exports[`should render error message if input is not valid 2`] = `"Please verify your input!"`;

exports[`should return 4 when calculating 1 + 3 1`] = `4`;
